### PR TITLE
FileLoadException: Add Profiler fix

### PR DESCRIPTION
**Pull request by zacharycmontoya**
### Issue 1
Our CLR Profiler currently assumes that as long as the `Datadog.Trace.ClrProfiler.Managed` assembly has been loaded domain-neutral once, all AppDomains can use that assembly. Under this false assumption, we always instrument domain-neutral assemblies which adds an AssemblyReference to our managed profiler and affects how the CLR shares code between AppDomains.

However, some applications may have BindingRedirect's specified on dependencies of the managed profiler like `System.Net.Http` which can force `Datadog.Trace.ClrProfiler.Managed` to no longer be loaded domain-neutral. When we instrument a domain-neutral assembly with a non-domain-neutral `Datadog.Trace.ClrProfiler.Managed` in one AppDomain and then another AppDomain tries to use the instrumented domain-neutral assembly, the second AppDomain crashes with the FileLoadException.

### Issue 2
Our new mechanism for automatically instrumenting AspNet uses the CLR Profiler to intercept a method call inside `System.Web` to call into `Datadog.Trace.ClrProfiler.Managed`. `System.Web` is always loaded domain-neutral but, as noted above, there are circumstances that can cause our assembly to not be loaded domain-neutral. We need to change the way we invoke the AspNet integration so that the assembly that runs the integration code is always domain-neutral.

### Addressing Issue 1
- Modify the profiler to never inject method calls to `Datadog.Trace.ClrProfiler.Managed` if the calling assembly was loaded domain-neutral.

### Addressing Issue 2
- Move basic types like `InterceptMethodAttribute` and friends to a new `Datadog.Trace.ClrProfiler.Managed.Core` that has no external dependencies.
- Move the AspNet integration into a new assembly called `Datadog.Trace.AspNet.Loader` whose only dependencies are `Datadog.Trace.ClrProfiler.Managed.Core` and `System.Web`. This assembly closure guarantees that the assembly will always be loaded domain-neutral. This is now a safe entry point for `System.Web` to call into during application setup. Also, the method responsible for adding the AspNet integration does not wrap any methods (that would require `Sigil`).
- To enable the minimal method call above that is simply called and returns to the caller without wrapping any target method, modify the method replacement infrastructure to support an `action`. The default value is `ReplaceTargetMethod` which performs the usual method replacement as before, but a new action `InsertFirst` will add a call to the method as the first instruction in the matching caller method. When generating `integrations.json`, add checks on the new `InsertFirst` integration methods.

## Consequences
- Fix `Loading this assembly would produce a different grant set from other instances. (Exception from HRESULT: 0x80131401)`
- For ASP.NET applications (or other scenarios where `LoaderOptimizationAttribute=MultiDomainHost`), there may be missing spans from calls between framework assemblies. Affected integrations include:
  - HttpMessageHandler
  - WebRequest
  - Wcf (server)

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs [hunk 1]
```diff
@@ -115,13 +109,7 @@ public static object GetResponse(object webRequest, int opCode, int mdToken, lon
...
         /// <returns>Returns the value returned by the inner method call.</returns>
         [InterceptMethod(
-            TargetAssembly = "System", // .NET Framework
-            TargetType = WebRequestTypeName,
-            TargetSignatureTypes = new[] { "System.Threading.Tasks.Task`1<System.Net.WebResponse>" },
-            TargetMinimumVersion = Major4,
-            TargetMaximumVersion = Major4)]
-        [InterceptMethod(
-            TargetAssembly = "System.Net.Requests", // .NET Core
+            TargetAssemblies = new[] { "System", "System.Net.Requests" },
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Should we split these additions into another PR?

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 2]
```diff
@@ -240,19 +232,38 @@ HRESULT STDMETHODCALLTYPE CorProfiler::AssemblyLoadFinished(AssemblyID assembly_
...
+        }
+        // Set the managed_profiler_unsafe_to_instrument_domain_neutral flag whenever the profiler is not loaded shared
+        else {
+          Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was not loaded domain-neutral");
+          managed_profiler_unsafe_to_instrument_domain_neutral = true;
+
+          // If the profiler was previously loaded domain-neutral, emit a warning pointing out this problematic failure mode
+          if (managed_profiler_loaded_domain_neutral) {
+            Warn("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was NOT loaded domain-neutral. The CLR determined that one of the dependencies of Datadog.Trace.ClrProfiler.Managed was unable to load in the AppDomain \"EE Shared Assembly Repository\". Enable DD_TRACE_DEBUG to see this debug-level info.");
+            Warn("WARNING: The profiler in this process (in IIS this means the Application Pool) will stop instrumenting Framework assemblies in the GAC such as System.Data, System.Net.Http, etc. To fix this issue, ensure that all BindingRedirects for System.Net.Http and dependencies of Datadog.Trace.ClrProfiler.Managed result in assemblies found in the GAC.");
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Great messages here Zach.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 3]
```diff
@@ -780,6 +789,27 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
...
+      // At this point we know we've hit a match. Error out if
+      //   1) The target assembly is Datadog.Trace.ClrProfiler.Managed
+      //   2) The calling assembly is domain-neutral
+      //   3) Datadog.Trace.ClrProfiler.Managed was never loaded domain-neutral
+      //        or it more recently has been loaded in a named AppDomain
+      // is domain-neutral but Datadog.Trace.ClrProfiler.Managed is
+      if (method_replacement.wrapper_method.assembly.name == "Datadog.Trace.ClrProfiler.Managed"_W &&
+          runtime_information_.is_desktop() && corlib_module_loaded &&
+          module_metadata->app_domain_id == corlib_app_domain_id &&
+          (!managed_profiler_loaded_domain_neutral || managed_profiler_unsafe_to_instrument_domain_neutral)) {
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Interesting that you added this: `method_replacement.wrapper_method.assembly.name == "Datadog.Trace.ClrProfiler.Managed"_W`
Is that necessary because of the `InterceptMethod` reference in the Loader assembly?

Might be a micro-optimization, but I think that's important at this level: Would you be amenable to re-ordering these conditions so that the string comparison is last?
Do you think that would make much of a difference here?

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs (hunk 1 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Sure, I can split that out.

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/WebRequestIntegration.cs]: Done in 0f6f8dd3bc3ffe32752f6e884b195dc656b2d86e

File: src/Datadog.Trace.AspNet.Loader/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        private static bool httpModuleRegistered = false;
+
+        /// <summary>
+        /// Injects a call to load the AspNetModule
+        /// </summary>
+        [InterceptMethod(
+            Integration = "AspNet",
+            CallerAssembly = "System.Web",
+            CallerType = "System.Web.Compilation.BuildManager",
+            CallerMethod = "InvokePreStartInitMethodsCore",
```
Comment (bobuva) [inline src/Datadog.Trace.AspNet.Loader/Startup.cs]: Was there a difference in intercepting at this point instead of in `InvokePreStartInitMethods`?

Comment (zacharycmontoya) [reply src/Datadog.Trace.AspNet.Loader/Startup.cs]: Hmm, looking back at this decision, I don't think there's any difference between the two callsites. I think I ended up choosing this inner callsite based off a different solution I was testing but didn't end up using.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 3 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Good idea. Yes, I will re-order the conditions so the string comparison is done last.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 2 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Thanks, but I sadly ended up stripping most of this detailed text because it no longer applies, as the decision now is to just never instrument domain-neutral assemblies, instead of basing the decision off the managed profiler

File: src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    /// Enum that instructs the CLR profiler where to insert the method call
+    /// in the caller's method body.
+    /// </summary>
+    public enum MethodReplacementActionType
+    {
+        /// <summary>
+        /// The new method call should replace the original method call in the
+        /// caller's body.
+        /// </summary>
+        ReplaceTargetMethod,
```
[the remark this answers is not in the document]
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: Are the new comments good, or is there anything I should re-word?

File: src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+        /// <summary>
+        /// The new method call should be placed at the beginning of the caller's body.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: Are the new comments good, or is there anything I should re-word?

[the remark this answers is not in the document]
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: Yes, good stuff! Thanks.

File: src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: Yes, as said above.

File: src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <RootNamespace>Datadog.Trace.ClrProfiler</RootNamespace>
+
+    <!-- NuGet -->
+    <Version>1.14.0</Version>
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj]: Should we keep the assembly version semi-constant (`1.0.0`) like we do for `Datadog.Trace.AspNet`?

File: tools/PrepareRelease/SetAllVersions.cs [hunk 8]
```diff
@@ -48,6 +48,10 @@ public static void Run()
                 "src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj",
                 NugetVersionReplace);
 
+            SynchronizeVersion(
+                "src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj",
```
Comment (lucaspimentel) [inline tools/PrepareRelease/SetAllVersions.cs]: We'll want to remove this one if we make the assembly version constant.

File: src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj (hunk 7 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj]: Good idea, I'll change that

File: src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+namespace Datadog.Trace.ClrProfiler
+{
+    /// <summary>
+    /// Enum that instructs the CLR profiler, during JIT compilation of a method,
+    /// where to insert a method call to the intercept method.
+    /// </summary>
+    public enum MethodReplacementActionType
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: Can we make this `internal`?

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/MethodReplacementActionType.cs]: If we set this to `internal` we won't be able to set the attribute when declaring the integration method. Since assembly can't be compiled against anyways (we don't provide it in a NuGet package), I think this should be fine as-is

File: src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj (hunk 7 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed.Core/Datadog.Trace.ClrProfiler.Managed.Core.csproj]: Fixed in d3c5fd865a5332058a9ff2f50ce2242164354550
